src/python/pants/option/parser.py

```diff
@@ -503,6 +505,38 @@ def error(
         if not is_enum and member_type not in self._allowed_member_types:
             error(InvalidMemberType, member_type=member_type.__name__)
 
+        # check type of default value
+        default_value = kwargs.get("default")
+        if default_value is not None:
+            if isinstance(default_value, str) and type_arg != str:
+                # attempt to parse default value, for correctness..
+                # custom function types will have to implement their own validation
+                default_value = self.to_value_type(default_value, type_arg, member_type, "")
+                if hasattr(default_value, "val"):
+                    default_value = default_value.val
```

Comment:
It seems like sufficient validation to have successfully converted the `default_value` `to_value_type`, without proceeding to the `if` statements below. Is adjusting `default_value` and falling through here intended to capture a particular case? If so, would be good to explain that in the comment above.

Reply:
indeed, the idea is to verify that a custom type does not bonker up by returning some other type than expected.